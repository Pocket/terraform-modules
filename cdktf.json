{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
