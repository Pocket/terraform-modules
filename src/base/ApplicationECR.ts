import { Resource } from 'cdktf';
import { Construct } from 'constructs';
import {
  EcrRepository,
  EcrRepositoryConfig,
  EcrLifecyclePolicy,
  EcrLifecyclePolicyConfig,
} from '../../.gen/providers/aws';

export interface ECRProps {
  name: string;
  tags?: { [key: string]: string };
}

export class ApplicationECR extends Resource {
  public readonly repo;

  constructor(scope: Construct, name: string, config: ECRProps) {
    super(scope, name);

    const ecrConfig: EcrRepositoryConfig = {
      name: config.name,
      tags: config.tags,
      imageScanningConfiguration: [
        {
          scanOnPush: true, // scans docker image for vulnerabilities
        },
      ],
    };

    this.repo = new EcrRepository(this, 'ecr-repo', ecrConfig);

    // this is our default policy
    // perhaps this should be defined elsewhere? or allow to be overwritten?
    // decisions for another day...
    const policy = {
      rules: [
        {
          rulePriority: 1,
          description: 'expire old images',
          selection: {
            tagStatus: 'any',
            countType: 'imageCountMoreThan',
            countNumber: 800,
          },
          action: {
            type: 'expire',
          },
        },
      ],
    };

    const ecrPolicyConfig: EcrLifecyclePolicyConfig = {
      repository: config.name,
      policy: JSON.stringify(policy),
    };

    new EcrLifecyclePolicy(this, 'ecr-repo-lifecyclepolicy', ecrPolicyConfig);
  }
}
