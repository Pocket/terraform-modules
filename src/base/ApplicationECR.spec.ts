import { Testing, TerraformStack } from 'cdktf';
import { ApplicationECR } from './ApplicationECR';

test('renders an ECR without tags', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationECR(stack, 'testECR', {
    name: 'bowling',
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders an ECR with tags', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationECR(stack, 'testECR', {
    name: 'bowling',
    tags: {
      name: 'rug',
      description: 'tiedtheroomtogether',
    },
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});
