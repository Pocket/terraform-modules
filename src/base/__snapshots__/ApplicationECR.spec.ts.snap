// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an ECR with tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_ecr_repository\\": {
      \\"test_testECR_ecrrepo_CC5FEBA3\\": {
        \\"name\\": \\"bowling\\",
        \\"tags\\": {
          \\"name\\": \\"rug\\",
          \\"description\\": \\"tiedtheroomtogether\\"
        },
        \\"image_scanning_configuration\\": [
          {
            \\"scan_on_push\\": true
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECR/ecr-repo\\",
            \\"uniqueId\\": \\"test_testECR_ecrrepo_CC5FEBA3\\"
          }
        }
      }
    },
    \\"aws_ecr_lifecycle_policy\\": {
      \\"test_testECR_ecrrepolifecyclepolicy_03E252E4\\": {
        \\"policy\\": \\"{\\\\\\"rules\\\\\\":[{\\\\\\"rulePriority\\\\\\":1,\\\\\\"description\\\\\\":\\\\\\"expire old images\\\\\\",\\\\\\"selection\\\\\\":{\\\\\\"tagStatus\\\\\\":\\\\\\"any\\\\\\",\\\\\\"countType\\\\\\":\\\\\\"imageCountMoreThan\\\\\\",\\\\\\"countNumber\\\\\\":800},\\\\\\"action\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"expire\\\\\\"}}]}\\",
        \\"repository\\": \\"bowling\\",
        \\"depends_on\\": [
          \\"aws_ecr_repository.test_testECR_ecrrepo_CC5FEBA3\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECR/ecr-repo-lifecyclepolicy\\",
            \\"uniqueId\\": \\"test_testECR_ecrrepolifecyclepolicy_03E252E4\\"
          }
        }
      }
    }
  }
}"
`;

exports[`renders an ECR without tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_ecr_repository\\": {
      \\"test_testECR_ecrrepo_CC5FEBA3\\": {
        \\"name\\": \\"bowling\\",
        \\"image_scanning_configuration\\": [
          {
            \\"scan_on_push\\": true
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECR/ecr-repo\\",
            \\"uniqueId\\": \\"test_testECR_ecrrepo_CC5FEBA3\\"
          }
        }
      }
    },
    \\"aws_ecr_lifecycle_policy\\": {
      \\"test_testECR_ecrrepolifecyclepolicy_03E252E4\\": {
        \\"policy\\": \\"{\\\\\\"rules\\\\\\":[{\\\\\\"rulePriority\\\\\\":1,\\\\\\"description\\\\\\":\\\\\\"expire old images\\\\\\",\\\\\\"selection\\\\\\":{\\\\\\"tagStatus\\\\\\":\\\\\\"any\\\\\\",\\\\\\"countType\\\\\\":\\\\\\"imageCountMoreThan\\\\\\",\\\\\\"countNumber\\\\\\":800},\\\\\\"action\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"expire\\\\\\"}}]}\\",
        \\"repository\\": \\"bowling\\",
        \\"depends_on\\": [
          \\"aws_ecr_repository.test_testECR_ecrrepo_CC5FEBA3\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECR/ecr-repo-lifecyclepolicy\\",
            \\"uniqueId\\": \\"test_testECR_ecrrepolifecyclepolicy_03E252E4\\"
          }
        }
      }
    }
  }
}"
`;
